### PR TITLE
Fix rollback of TxMeta and Certs

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#1575 

# Overview

- [x] I have constructed a regression test using the DB and a mock chain.
- [x] I tweaked the `rollbackTo` code to clear from `nearestPoint` instead of `point`. I.e. the actual rollback point instead of the requested rollback point.
- [x] Improved state machine counterexamples slightly

# Comments

- We should perhaps think about why no higher-level property caught this bug previously.

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 1]
```diff
@@ -10,7 +10,7 @@
 {-# LANGUAGE TupleSections #-}
 {-# LANGUAGE TypeApplications #-}
 
-{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -fno-warn-orphans -fno-warn-unused-imports #-}
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: Oops

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs [hunk 2]
```diff
@@ -505,21 +505,21 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
                     Just _  -> lift $ throwIO (ErrNoOlderCheckpoint wid point)
                 Just nearestPoint -> do
                     deleteCheckpoints wid
-                        [ CheckpointSlot >. point
+                        [ CheckpointSlot >. nearestPoint
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/DB/Sqlite.hs:494]: 🤔 Should work, but the change should also not be necessary.

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs [hunk 3]
```diff
@@ -505,21 +505,21 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
...
                         ]
                     deleteDelegationCertificates wid
                         [ CertSlot >. point
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/DB/Sqlite.hs]: Should also be `nearestPoint`

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs (hunk 2 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/DB/Sqlite.hs:494]: Keeping it for consistency.

**Comment by Anviking** [comment]
State machine tests will fail?

bors try

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 4]
```diff
@@ -471,6 +484,88 @@ fileModeSpec =  do
...
+                -- Since we increment the blockHeight faster than naturally,
+                -- The wallet should have pruned everything but the first and last
+                -- checkpoint.
+                --
+                -- So when we rollback to the previous step...
+                atomically . void . unsafeRunExceptT $
+                    rollbackTo testPk (SlotId 2 0)
+                Just cp <- atomically $ readCheckpoint testPk
+
+                -- We end up at genesis.
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: That comment is superfluous. Comments should give detail about "why" , not "what". What is written just below and already perfectly readable :ok_hand:

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 5]
```diff
@@ -471,6 +484,88 @@ fileModeSpec =  do
...
+                slotId (currentTip cp) `shouldBe` (SlotId 0 0)
+
+                -- (Outgoing, 2) should /not/ be a "InLedger" transaction!
+                --
+                -- Whether is is now pending, or removed completely is not the
+                -- concern of this test.
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: Same as above, the comment is superfluous.

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 6]
```diff
@@ -471,6 +484,88 @@ fileModeSpec =  do
...
+                        Just cpA <- atomically $ readCheckpoint testPk
+                        let slotA = slotId $ currentTip cpA
+                        let Quantity bhA = blockHeight $ currentTip cpA
+                        let hashA = headerHash $ currentTip cpA
+                        let fakeBlock = Block
+                                (BlockHeader
+                                { slotId = slotA { epochNumber = (epochNumber slotA) + 1}
+                                -- Increment blockHeight by steps greater than k
+                                -- Such that old checkpoints are always pruned.
+                                , blockHeight = Quantity $ bhA + 5000
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:495]: I am not sure what the pruning has to do with this test and why it matters to introduce this extra complexity in the test?

**Review by KtorZ — COMMENTED** [review]
Looks good, although I wonder why not writing a simple property here using arbitrary data rather using mock and fake and dummy everywhere :thinking:  ?

Also, seems like `prop_rollbackTxHistory`, for a good reason it seems. The test isn't actually taking care of any consistency consideration between checkpoints and history. 

Also, the database model needs to be updated accordingly, hence the state-machine failures.

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs (hunk 4 above)
Comment (Anviking) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: > That comment is superfluous

Yes, perhaps. I totally agree in general.

My reasoning might have been: There is a lot of boilerplate here and "What"-comments highlight the important stuff.

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs (hunk 6 above)
Comment (Anviking) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:495]: Hm… indeed

Comment (Anviking) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:495]: No, this is essential. Without pruning there is no problem.

If a checkpoint exists for every slot, then `nearestPoint` and `requestedPoint` are going to be equal, and the difference does not matter (Also see diagram in #1575)

File: lib/core/test/unit/Cardano/Wallet/DB/Properties.hs [hunk 7]
```diff
@@ -726,10 +726,13 @@ prop_rollbackTxHistory db@DBLayer{..} (InitialCheckpoint cp0) (GenTxHistory txs0
...
             unsafeRunExceptT $ putTxHistory wid txs0
 
-    prop wid point = do
-        _ <- run $ unsafeRunExceptT $ mapExceptT atomically $ rollbackTo wid point
+    prop wid requestedPoint = do
+        point <- run $ unsafeRunExceptT $ mapExceptT atomically $ rollbackTo wid requestedPoint
         txs <- run $ atomically $ readTxHistory wid Descending wholeRange Nothing
-        monitor $ counterexample $ "\nTx history after rollback: \n" <> fmt txs
+
+        monitor $ counterexample $ "\n" <> "Actual Rollback Point:\n" <> (pretty point)
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/Wallet/DB/Properties.hs:734]: This is weird. 
```
         Requested Rollback point:
         79.0

         Actual Rollback Point:
         0.0

         Original tx history:
         -  9eb44fb2 14.6 in ledger outgoing 2

         New tx history:
         -  9eb44fb2 14.6 in ledger outgoing 2
         Outgoing txs are reschuled ✗

  To rerun use: --match "/Cardano.Wallet.DB.MVar/MVar/rollback/Correctly re-construct tx history on rollbacks/"
```

Through experimentation it seems the `TxMetaSlot >. nearestPoint` predicate is ineffective. (If removed all outgoing become pending, all incoming are removed). But still: wat?

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs (hunk 4 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: Functions should be used then, they are clearer than comments, and re-usable.

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs (hunk 6 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:495]: But you don't have to insert a checkpoint at every slot for the test in the first place?

File: lib/core/test/unit/Cardano/Wallet/DB/Properties.hs (hunk 7 above)
Comment (Anviking) [reply lib/core/test/unit/Cardano/Wallet/DB/Properties.hs:734]: Not sure what fixed it 🤷‍♂️

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs (hunk 6 above)
Comment (Anviking) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:495]: I wanted to include [`prune`](https://github.com/input-output-hk/cardano-wallet/blob/7e0aa90186462721397b22a3ca90967862716fbb/lib/core/src/Cardano/Wallet.hs#L767) similar to `restoreBlocks`, to convince myself that the bug could happen in reality.

But, yes, we could also specify which checkpoints to write manually.

**Comment by Anviking** [comment]
Could move
- Make (Wallet s) a plain ADT to derive Generic 
- Add more shrinking to State Machine Tests
- Derive ToExpr instances using generic

to separate PR if desired.

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs (hunk 4 above)
Comment (Anviking) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: 755065b

File: lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 8]
```diff
@@ -413,6 +414,14 @@ instance Arbitrary UTxO where
             <*> vector n
         return $ UTxO $ Map.fromList u
 
+
+instance Arbitrary a => Arbitrary (Range a) where
+    arbitrary = Range <$> arbitrary <*> arbitrary
+
+    shrink (Range from to) =
+        [Range from' to | from' <- shrink from]
+        ++ [Range from to' | to' <- shrink to]
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:424]: Are we sure that `to'` is necessarily bigger than `from` here ? For `from'`, it should shrink to a value smaller than `from` so there's no obvious problem, but for `to'` :shrug:

**Review by KtorZ — APPROVED** [review]
LGTM. I wonder how this previously property got fixed though...

**Pull request merged**